sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SdkSpanBackend.scala

```diff
@@ -0,0 +1,311 @@
+  *
+  * @tparam F
+  *   the higher-kinded type of a polymorphic effect
+  */
+private[trace] final class SdkSpanBackend[F[_]: Monad: Clock: Console] private (
+    spanProcessor: SpanProcessor[F],
+    immutableState: SdkSpanBackend.ImmutableState,
+    mutableState: AtomicCell[F, SdkSpanBackend.MutableState]
+) extends Span.Backend[F]
+    with SpanRef[F] {
+
+  def meta: InstrumentMeta[F] =
+    InstrumentMeta.enabled
+
+  def context: SpanContext =
+    immutableState.context
+
+  def updateName(name: String): F[Unit] =
+    updateState("updateName")(_.copy(name = name)).void
+
+  def addAttributes(attributes: Attribute[_]*): F[Unit] =
+    updateState("addAttributes") { s =>
+      val next = Attributes.newBuilder
+        .addAll(s.attributes)
+        .addAll(attributes)
+        .result()
+
+      s.copy(attributes = next)
+    }.unlessA(attributes.isEmpty)
+
+  def addEvent(name: String, attributes: Attribute[_]*): F[Unit] =
+    for {
+      now <- Clock[F].realTime
+      _ <- addEvent(name, now, attributes: _*)
+    } yield ()
+
+  def addEvent(
+      name: String,
+      timestamp: FiniteDuration,
+      attributes: Attribute[_]*
+  ): F[Unit] =
+    addTimedEvent(
+      EventData(name, timestamp, Attributes.fromSpecific(attributes))
+    )
+
+  def recordException(
+      exception: Throwable,
+      attributes: Attribute[_]*
+  ): F[Unit] =
+    for {
+      now <- Clock[F].realTime
+      _ <- addTimedEvent(
+        EventData.fromException(
+          now,
+          exception,
+          Attributes.fromSpecific(attributes),
+          false
+        )
+      )
+    } yield ()
+
+  def setStatus(status: Status): F[Unit] =
+    updateState("setStatus") { s =>
+      s.copy(status = StatusData(status))
+    }.void
+
+  def setStatus(status: Status, description: String): F[Unit] =
+    updateState("setStatus") { s =>
+      s.copy(status = StatusData(status, description))
+    }.void
+
+  private[otel4s] def end: F[Unit] =
+    for {
+      now <- Clock[F].realTime
+      _ <- end(now)
+    } yield ()
+
+  private[otel4s] def end(timestamp: FiniteDuration): F[Unit] = {
+    for {
+      updated <- updateState("end")(s => s.copy(endTimestamp = Some(timestamp)))
+      _ <- toSpanData.flatMap(span => spanProcessor.onEnd(span)).whenA(updated)
+    } yield ()
+  }
+
+  private def addTimedEvent(event: EventData): F[Unit] =
+    updateState("addEvent")(s => s.copy(events = s.events :+ event)).void
+
+  // applies modifications while the span is still active
+  // modifications are ignored when the span is ended
+  private def updateState(
+      method: String
+  )(update: MutableState => MutableState): F[Boolean] =
+    mutableState
+      .modify { state =>
+        if (state.endTimestamp.isDefined) (state, false)
+        else (update(state), true)
+      }
+      .flatTap { modified =>
+        Console[F]
+          .println(
+            s"SdkSpanBackend: calling [$method] on the ended span $context"
+          )
```

Comment:
Since `Console` is not a real logger and a lot of users might just pass the default, is this annoying? Or was it just for debugging?

Reply:
It shouldn't be logged often. The log message may occur if you try to update the span that has already ended.

For example:
```scala
for {
  escapedSpan <- Tracer[F].span("my-span").use { span =>
    IO.pure(span)
  }
  _ <- escapedSpan.updateName("new name") // <- 'SdkSpanBackend: calling [updateName] on the ended span ...' will be logged
} yield ()
```

Reply:
Hmm, right. But it doesn't serious enough to be an `errorln` I suppose.

Reply:
We can hide these warnings behind a config (e.g. `traceProvider.enableConsoleWarnings`).

Reply:
> Hmm, right. But it doesn't serious enough to be an `errorln` I suppose.

yeah. If we had a proper logging system, I would use `warn`.